### DISCUSSION TITLE
Wait only if Response.Code is 429

### DISCUSSION THREAD

**Discussion by gcuadrado** [abuzuhri/Amazon-SP-API-CSharp#61] [Ideas]
At the moment, if we have the IsActiveLimitRate set to true, after every request, we have to wait a while based on x-amzn-RateLimit-Limit value. 

I think it's should be better if we don't wait by default. We should only wait if a request fails with 429 code, in this case, we should wait and then retry the request.

What do you think about it?

**Comment by gcuadrado** [comment]
Something like this:
`protected T ExecuteRequest<T>() where T : new()`

        {
            Request = TokenGeneration.SignWithSTSKeysAndSecurityToken(Request, RequestClient.BaseUrl.Host, AmazonCredential);
            var response = RequestClient.Execute<T>(Request);

            if (response.StatusCode == 429)
            {
                SleepForRateLimit(response.Headers);
                response = RequestClient.Execute<T>(Request);
            } 
            ParseResponse(response);

            if (response.StatusCode==HttpStatusCode.OK && !string.IsNullOrEmpty(response.Content) && response.Data == null)
            {
                response.Data = JsonConvert.DeserializeObject<T>(response.Content);
            }
            return response.Data;
        }

**Comment by abuzuhri** [comment]
Hello @gcuadrado 

As i know if get error 429 then you will get the true ratelimit , do you test this code ?

**Comment by gcuadrado** [comment]
Hi @abuzuhri it is not tested, it is just an idea.

I don't understand what do you mean with "true rate limit". (Maybe I have to read again this part of Amazon documentation)

If you get 429 you won't get the result for your request, but if you wait and repeat the request, you will get the correct result

**Comment by abuzuhri** [comment]
Hi @gcuadrado 

What I mean is

If there is exception (429) happen then no rate limit return at header ?
you need to wait the default rate limit that is defined in amazon doc

**Reply by gcuadrado** [reply]
Ah ok, I will test it tomorrow to check if that header is present when get the 429.

Thank you

**Comment by gcuadrado** [comment]
I have been testing.

You were right, if you reach the limit and get the 429, you won't have the rate limit header.

Maybe we can add a sleepTime property to AmazonConnection class and pass it as reference to all their services classes that inhertis from RequestService.

When execute a request, if it's correct, we update the sleepTime value based on rate limit header, if it's fails, we wait the time specified in sleepTime property.

With this way we could share sleepTime across all services, the cons are that sleepTime would be different for each AmazonConnection instance, but in normal conditions, we should have only one.

**Comment by abuzuhri** [comment]
I will try next week to work in best solution to solve this problem , also if you can with good solution we can speak about it i will take look to other library and see how they solve it

**Reply by gcuadrado** [reply]
Have you done any progress in this? Probably I must work in any improvement next week.

**Reply by abuzuhri** [reply]
No not yet , I was busy with finish reports .

Let think in the best solution and how others library solve the problem
